src/librustc_mir/transform/unreachable_prop.rs

```diff
@@ -0,0 +1,108 @@
+//! A pass that propagates the unreachable terminator of a block to its predecessors
+//! when all of their successors are unreachable. This is achieved through a
+//! post-order traversal of the blocks.
+
+use crate::transform::simplify;
+use crate::transform::{MirPass, MirSource};
+use rustc::mir::*;
+use rustc::ty::TyCtxt;
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+use std::borrow::Cow;
+
+pub struct UnreachablePropagation;
+
+impl MirPass<'_> for UnreachablePropagation {
+    fn run_pass<'tcx>(
+        &self,
+        _tcx: TyCtxt<'tcx>,
+        _: MirSource<'tcx>,
+        body: &mut BodyAndCache<'tcx>,
+    ) {
+        let mut unreachable_blocks = FxHashSet::default();
+        let mut replacements = FxHashMap::default();
+        let is_asm_stmt = |stmt: &Statement<'_>| match stmt.kind {
+            StatementKind::InlineAsm(..) => true,
+            _ => false,
+        };
+
+        for (bb, bb_data) in traversal::postorder(body) {
+            let terminator = bb_data.terminator();
+
+            if let TerminatorKind::Unreachable = terminator.kind {
+                unreachable_blocks.insert(bb);
+            } else {
```

Comment:
Maybe check for inline asm first?

Reply:
I believe you are right because if we have
```
bb0 {
    ...
    goto -> bb1;
}
bb1 {
     <diverging asm stmt>
     unreachable;
}
```
we cannot propagate unreachable to bb0, correct? Basically it should not matter if the asm instruction is in bb0 or in bb1 in this example. Just confirm so I fix this.

Reply:
@oli-obk Your opinion on this?

Reply:
if the asm instruction is in `bb0`, then you can propagate the unreachable from bb1 to bb0 (if bb1 doesn't also have asm instructions), but in the example you gave, no change should be done to the MIR